great_expectations/datasource/datasource_dict.py

```diff
@@ -170,20 +167,26 @@ def __contains__(self, name: object) -> bool:
 
     @override
     def set_datasource(self, name: str, ds: FluentDatasource) -> FluentDatasource | None:
-        self.data[name] = ds
+        self.data[name] = self._add_ids(ds)
+        return ds
+
+    def _add_ids(self, ds: FluentDatasource) -> FluentDatasource:
+        # File and ephemeral contexts do not use the store, so we need to add IDs here.
+        # Note that this is used for both `add` and `update` operations.
```

Comment:
We should have ids when we update going forward but we need to do this on update now since we may have configs stored without ids?

Reply:
If I update a datasource, it may already have an id but any newly added children or grandchildren may not

Reply:
That makes sense, thanks!